src/sql/src/plan/query.rs

```diff
@@ -2968,24 +2966,32 @@ fn plan_case<'a>(
 fn plan_literal<'a>(l: &'a Value) -> Result<CoercibleScalarExpr, anyhow::Error> {
     let (datum, scalar_type) = match l {
         Value::Number(s) => {
-            let d: Decimal = s.parse()?;
-            if d.scale() == 0 {
-                let significand = d.significand();
-                if let Ok(n) = significand.try_into() {
-                    (Datum::Int32(n), ScalarType::Int32)
-                } else if let Ok(n) = significand.try_into() {
-                    (Datum::Int64(n), ScalarType::Int64)
-                } else {
-                    (
-                        Datum::from(significand),
-                        ScalarType::Decimal(MAX_DECIMAL_PRECISION, d.scale()),
-                    )
+            match strconv::parse_numeric(s.as_str()) {
+                Ok(d) => {
+                    if !s.contains(&['E', '.'][..]) {
+                        // Maybe representable as an int?
+                        if let Ok(n) = d.0.try_into() {
+                            (Datum::Int32(n), ScalarType::Int32)
+                        } else if let Ok(n) = d.0.try_into() {
+                            (Datum::Int64(n), ScalarType::Int64)
+                        } else {
+                            (Datum::Numeric(d), ScalarType::Numeric { scale: None })
+                        }
+                    } else {
+                        (Datum::Numeric(d), ScalarType::Numeric { scale: None })
+                    }
+                }
+                Err(..) => {
+                    let f64_val = s.parse::<f64>()?;
+                    if f64_val.is_finite() {
+                        (Datum::from(f64_val), ScalarType::Float64)
+                    } else {
+                        bail!(
+                            "number {} exceeds min/max representation of all numeric types",
+                            s
+                        )
+                    }
```

Comment:
This gives me some anxiety because AFAIK PostgreSQL would never do this, right? Does anything break if we just hard reject numeric literals that don't fit in `numeric`?

Reply:
Yeah, PG would never do this because their decimals' precision exceeds `f64`'s. Glad to change this to just error on literals that exceed 39 DOP, as you suggest.